### DISCUSSION TITLE
Empty Optional Parameters And `alphanum`

### DISCUSSION THREAD

**Discussion by smatric** [go-playground/validator#718] [Q&A]
My endpoint accepts two optional parameters:

```
type Params struct {
	FYear  int    `form:"f-year" validate:"gt=1800,lt=2100"`
	FQuery string `form:"f-query" validate:"alphanum"`
}
```

However, validation fails on the field tagged as `alphanum`:

```
Error:Field validation for 'FQuery' failed on the 'alphanum' tag
```

 How can I make the rule to be applied only when the field has a value?

**Comment by smatric** [comment]
Gin framework automatically assigns zero values to the fields corresponding to query string parameters that are missing. If I understand this correctly, this is why I cannot have an optional URL parameter and use validation rules such as "gt=1800" in the example above (because the value would be `0` if omitted). 

If you could add support for a library such as [gopkg.in/guregu/null.v4](https://github.com/guregu/null/tree/v4.0.0), it would solve the problem. I could define fields as nullable:

```
type Params struct {
	FYear  null.Int    `form:"f-year" validate:"gt=1800,lt=2100"`
	FQuery null.String `form:"f-query" validate:"alphanum"`
}
```
and, at the same time, apply validation that wouldn't be affected by zero values.

Please consider this feature request.

**Comment by deankarn** [comment]
Hey @smatric there are actually two ways to do what you want today :)

1. Use the `omitempty` tag and when the value is empty validator we skip all subsequent validations.
2. Register a `CustomTypeFunc` (Custom Type Function) to handle just like the **sql.Null*** types [here](https://github.com/go-playground/validator/blob/b926bf0df97108effd04f9325365cd7969935e16/validator_test.go#L2234) in the tests.

let me know if this helps.

**Reply by smatric** [reply]
I somehow overlooked that. Thank you!
